File: packages/crypto/src/aead.ts
/**
 * Authenticated encryption.
 *
 * XChaCha20-Poly1305, with a 24-byte nonce. The extended nonce is the reason for the choice: at
 * 24 bytes a randomly generated nonce has no meaningful collision risk, so nonces can be random
 * per message and no component has to maintain a counter that must never repeat. AES-GCM's
 * 12-byte nonce would force exactly that bookkeeping, and nonce reuse under GCM is catastrophic
 * — it leaks the authentication key, not just one message.
 *
 * ChaCha20 is also the right shape for the target device: a software stream cipher, fast and
 * constant-time without hardware AES, which many of the cheap Android phones Migo targets do
 * not have.
 *
 * Associated data is always supplied and always authenticated. For a ratchet message it is the
 * header; for a stored blob it is the record identity. This is what stops a valid ciphertext
 * from being replayed into a different context.
 *
 * ## Why not WebCrypto
 *
 * `crypto.subtle` has no ChaCha20 at all — it offers AES-GCM, whose 12-byte nonce is the
 * bookkeeping problem above — and every one of its operations is a `Promise`. The ratchet runs
 * inside a synchronous decode path, so an async AEAD would force the entire message pipeline to
 * be async for no security benefit. `@noble/ciphers` is audited, has no dependencies, and is
 * synchronous, which is why ADR-0003 names it. WebCrypto is still used for one thing: the
 * random bytes, in {@link SymmetricKey.generate} and {@link seal}, because a CSPRNG is exactly
 * the primitive that should come from the platform.
 */

import { xchacha20poly1305 } from '@noble/ciphers/chacha.js';

import { CryptoError } from './errors.js';

/** Symmetric key length in bytes. */
export const KEY_LEN = 32;
/** Nonce length in bytes. */
export const NONCE_LEN = 24;
/** Poly1305 authentication tag length in bytes. */
export const TAG_LEN = 16;

/**
 * Fills `out` with cryptographically secure random bytes.
 *
 * Unlike the Rust side, the generator is not an injectable parameter. There, `seal` takes
 * `&mut dyn Random` so that a caller cannot reach for the seeded generator used in
 * deterministic simulation; in JavaScript the equivalent hazard is a test helper that returns
 * `Math.random()` bytes and a reviewer who does not notice. Making the source unreachable and
 * routing determinism through {@link sealWithNonce} instead removes the parameter that could
 * carry the mistake.
 */
function randomBytes(out: Uint8Array): void {
  const source = globalThis.crypto;
  if (typeof source?.getRandomValues !== 'function') {
    // Refusing is the only safe answer. A fallback to a non-cryptographic generator would
    // produce keys that look random in every test and are predictable in production.
    throw new TypeError('no cryptographic random source: crypto.getRandomValues is unavailable');
  }
  source.getRandomValues(out);
}

/**
 * A 256-bit symmetric key.
 *
 * There is no `Display` and no useful default inspection: a key that can be printed eventually
 * is printed, into a log that is retained for ninety days. The bytes live in a `#private` field
 * and {@link toString}, {@link toJSON} and Node's inspect hook all answer `SymmetricKey(***)`.
 */
export class SymmetricKey {
  readonly #bytes: Uint8Array;
  #destroyed = false;

  private constructor(bytes: Uint8Array) {
    this.#bytes = bytes;
  }

  /**
   * Wraps existing key bytes, copying them.
   *
   * Rust has both `from_bytes([u8; KEY_LEN])` and `parse(&[u8])` because its type system can
   * express the fixed width; TypeScript cannot, so the two collapse into one function that
   * checks the length at runtime. The copy matters for the same reason as in
   * {@link MacKey.fromBytes}: the caller's buffer and this key must not be the same memory, or
   * clearing one silently clears the other.
   */
  static fromBytes(bytes: Uint8Array): SymmetricKey {
    if (bytes.length !== KEY_LEN) {
      throw CryptoError.badLength('symmetric key', KEY_LEN, bytes.length);
    }
    return new SymmetricKey(bytes.slice());
  }

  /** Generates a fresh key from the platform CSPRNG. */
  static generate(): SymmetricKey {
    const bytes = new Uint8Array(KEY_LEN);
    randomBytes(bytes);
    return new SymmetricKey(bytes);
  }

  /**
   * Borrows the raw bytes. The greppable audit point for key material leaving this type.
   *
   * Returns the live buffer, not a copy — a copy would defeat {@link destroy} by scattering
   * duplicates, and would hide from an auditor how many of them exist. Callers must not write
   * into it. The same name as Rust's `expose`, so that `rg 'expose\('` finds every site in both
   * languages at once.
   */
  expose(): Uint8Array {
    if (this.#destroyed) {
      throw new TypeError('SymmetricKey has been destroyed');
    }
    return this.#bytes;
  }

  /**
   * Clears the key bytes and makes every later use throw.
   *
   * Best-effort in the same way {@link MacKey.destroy} is: a garbage collector may have moved
   * this buffer and left a copy behind, and nothing in the language can reach it. What is
   * guaranteed is that the live buffer stops holding a key when the session ends.
   */
  destroy(): void {
    this.#bytes.fill(0);
    this.#destroyed = true;
  }

  /** `SymmetricKey(***)`. Never the key. */
  toString(): string {
    return 'SymmetricKey(***)';
  }

  /** `JSON.stringify` of anything holding a key must not produce the key. */
  toJSON(): string {
    return 'SymmetricKey(***)';
  }

  /** `console.log` and `util.inspect` in Node. */
  [Symbol.for('nodejs.util.inspect.custom')](): string {
    return 'SymmetricKey(***)';
  }
}

/**
 * Encrypts `plaintext`, returning `nonce || ciphertext || tag`.
 *
 * The nonce is prefixed rather than tracked separately because every caller needs it and a
 * caller that has to remember to store it separately eventually will not.
 */
export function seal(
  key: SymmetricKey,
  associatedData: Uint8Array,
  plaintext: Uint8Array,
): Uint8Array {
  const nonce = new Uint8Array(NONCE_LEN);
  randomBytes(nonce);
  return sealWithNonce(key, nonce, associatedData, plaintext);
}

/**
 * Encrypts with a caller-supplied nonce.
 *
 * Exists for test vectors and for the ratchet, which derives its nonce from the message key so
 * that both sides agree without transmitting it. Application code should call {@link seal} and
 * let the nonce be random.
 */
export function sealWithNonce(
  key: SymmetricKey,
  nonce: Uint8Array,
  associatedData: Uint8Array,
  plaintext: Uint8Array,
): Uint8Array {
  requireNonce(nonce);
  const cipher = xchacha20poly1305(key.expose(), nonce, associatedData);
  const body = cipher.encrypt(plaintext);
  const out = new Uint8Array(NONCE_LEN + body.length);
  out.set(nonce, 0);
  out.set(body, NONCE_LEN);
  return out;
}

/** Decrypts `nonce || ciphertext || tag`. */
export function open(
  key: SymmetricKey,
  associatedData: Uint8Array,
  sealed: Uint8Array,
): Uint8Array {
  if (sealed.length < NONCE_LEN + TAG_LEN) {
    // Too short to contain a nonce and a tag, so there is nothing to authenticate. Reported as
    // a length problem rather than a decryption failure because it says something about the
    // shape of the input, which the sender can already see, and not about the key.
    throw CryptoError.badLength('sealed message', NONCE_LEN + TAG_LEN, sealed.length);
  }
  return openWithNonce(
    key,
    sealed.subarray(0, NONCE_LEN),
    associatedData,
    sealed.subarray(NONCE_LEN),
  );
}

/** Decrypts with an explicit nonce, for ciphertext stored without its nonce. */
export function openWithNonce(
  key: SymmetricKey,
  nonce: Uint8Array,
  associatedData: Uint8Array,
  body: Uint8Array,
): Uint8Array {
  requireNonce(nonce);
  if (body.length < TAG_LEN) {
    throw CryptoError.badLength('sealed body', TAG_LEN, body.length);
  }
  const cipher = xchacha20poly1305(key.expose(), nonce, associatedData);
  try {
    return cipher.decrypt(body);
  } catch {
    // Every cause collapses to one error: wrong key, wrong nonce, edited ciphertext, edited
    // associated data. Distinguishing them is exactly the information an attacker wants — that
    // is the shape of every padding oracle — and the receiver's action is the same in all four
    // cases. The library's own message is discarded rather than wrapped, because it is written
    // for a developer debugging a call, and this call's input is a hostile peer's bytes.
    throw CryptoError.decryptionFailed();
  }
}

function requireNonce(nonce: Uint8Array): void {
  if (nonce.length !== NONCE_LEN) {
    throw CryptoError.badLength('nonce', NONCE_LEN, nonce.length);
  }
}

File: server/crates/migo-crypto/src/aead.rs
//! Authenticated encryption.
//!
//! XChaCha20-Poly1305, with a 24-byte nonce. The extended nonce is the reason for
//! the choice: at 24 bytes a randomly generated nonce has no meaningful collision
//! risk, so nonces can be random per message and no component has to maintain a
//! counter that must never repeat. AES-GCM's 12-byte nonce would force exactly
//! that bookkeeping, and nonce reuse under GCM is catastrophic — it leaks the
//! authentication key, not just one message.
//!
//! ChaCha20 is also the right shape for the target device: it is a software
//! stream cipher, fast and constant-time without hardware AES, which many of the
//! cheap Android phones Migo targets do not have.
//!
//! Associated data is always supplied and always authenticated. For a ratchet
//! message it is the header; for a stored blob it is the record identity. This is
//! what stops a valid ciphertext from being replayed into a different context.

use chacha20poly1305::aead::{Aead, KeyInit, Payload};
use chacha20poly1305::{XChaCha20Poly1305, XNonce};
use migo_core::Random;
use zeroize::Zeroize;

use crate::error::{CryptoError, Result};

/// Symmetric key length in bytes.
pub const KEY_LEN: usize = 32;
/// Nonce length in bytes.
pub const NONCE_LEN: usize = 24;
/// Poly1305 authentication tag length in bytes.
pub const TAG_LEN: usize = 16;

/// A 256-bit symmetric key that clears itself when dropped.
///
/// There is no `Display` and no plain `Debug` output: a key that can be printed
/// eventually is printed, into a log that is retained for ninety days.
#[derive(Clone, PartialEq, Eq, Zeroize)]
#[zeroize(drop)]
pub struct SymmetricKey([u8; KEY_LEN]);

impl SymmetricKey {
    /// Wraps existing key bytes.
    #[must_use]
    pub fn from_bytes(bytes: [u8; KEY_LEN]) -> Self {
        Self(bytes)
    }

    /// Generates a fresh key.
    ///
    /// Takes the source explicitly so a caller cannot accidentally reach for a
    /// seeded generator: `SeededRandom` exists for deterministic simulation and
    /// must never produce a real key.
    pub fn generate(random: &mut dyn Random) -> Self {
        let mut bytes = [0u8; KEY_LEN];
        random.fill_bytes(&mut bytes);
        Self(bytes)
    }

    /// Parses a key from a slice.
    pub fn parse(bytes: &[u8]) -> Result<Self> {
        let array: [u8; KEY_LEN] = bytes.try_into().map_err(|_| CryptoError::BadLength {
            what: "symmetric key",
            expected: KEY_LEN,
            actual: bytes.len(),
        })?;
        Ok(Self(array))
    }

    /// Borrows the raw bytes. The greppable audit point for key material leaving
    /// this type.
    #[must_use]
    pub fn expose(&self) -> &[u8; KEY_LEN] {
        &self.0
    }
}

impl core::fmt::Debug for SymmetricKey {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.write_str("SymmetricKey(***)")
    }
}

/// Encrypts `plaintext`, returning `nonce || ciphertext || tag`.
///
/// The nonce is prefixed rather than tracked separately because every caller
/// needs it and a caller that has to remember to store it separately eventually
/// will not.
pub fn seal(
    key: &SymmetricKey,
    associated_data: &[u8],
    plaintext: &[u8],
    random: &mut dyn Random,
) -> Result<Vec<u8>> {
    let mut nonce_bytes = [0u8; NONCE_LEN];
    random.fill_bytes(&mut nonce_bytes);
    seal_with_nonce(key, &nonce_bytes, associated_data, plaintext)
}

/// Encrypts with a caller-supplied nonce.
///
/// Exists for test vectors and for the ratchet, which derives its nonce from the
/// message key so that both sides agree without transmitting it. Application code
/// should call [`seal`] and let the nonce be random.
pub fn seal_with_nonce(
    key: &SymmetricKey,
    nonce: &[u8; NONCE_LEN],
    associated_data: &[u8],
    plaintext: &[u8],
) -> Result<Vec<u8>> {
    let cipher = XChaCha20Poly1305::new(key.expose().into());
    let ciphertext = cipher
        .encrypt(
            XNonce::from_slice(nonce),
            Payload {
                msg: plaintext,
                aad: associated_data,
            },
        )
        .map_err(|_| CryptoError::DecryptionFailed)?;
    let mut out = Vec::with_capacity(NONCE_LEN + ciphertext.len());
    out.extend_from_slice(nonce);
    out.extend_from_slice(&ciphertext);
    Ok(out)
}

/// Decrypts `nonce || ciphertext || tag`.
pub fn open(key: &SymmetricKey, associated_data: &[u8], sealed: &[u8]) -> Result<Vec<u8>> {
    if sealed.len() < NONCE_LEN + TAG_LEN {
        return Err(CryptoError::BadLength {
            what: "sealed message",
            expected: NONCE_LEN + TAG_LEN,
            actual: sealed.len(),
        });
    }
    let (nonce, body) = sealed.split_at(NONCE_LEN);
    let cipher = XChaCha20Poly1305::new(key.expose().into());
    cipher
        .decrypt(
            XNonce::from_slice(nonce),
            Payload {
                msg: body,
                aad: associated_data,
            },
        )
        .map_err(|_| CryptoError::DecryptionFailed)
}

/// Decrypts with an explicit nonce, for ciphertext stored without its nonce.
pub fn open_with_nonce(
    key: &SymmetricKey,
    nonce: &[u8; NONCE_LEN],
    associated_data: &[u8],
    ciphertext: &[u8],
) -> Result<Vec<u8>> {
    let cipher = XChaCha20Poly1305::new(key.expose().into());
    cipher
        .decrypt(
            XNonce::from_slice(nonce),
            Payload {
                msg: ciphertext,
                aad: associated_data,
            },
        )
        .map_err(|_| CryptoError::DecryptionFailed)
}

#[cfg(test)]
mod tests {
    use super::*;
    use migo_core::SeededRandom;

    fn key() -> SymmetricKey {
        SymmetricKey::from_bytes([7u8; KEY_LEN])
    }

    #[test]
    fn round_trips() {
        let mut random = SeededRandom::new(1234);
        let sealed = seal(&key(), b"context", b"pesan rahasia", &mut random).expect("seals");
        let opened = open(&key(), b"context", &sealed).expect("opens");
        assert_eq!(opened, b"pesan rahasia");
    }

    #[test]
    fn ciphertext_expands_by_exactly_the_nonce_and_tag() {
        let mut random = SeededRandom::new(1);
        let sealed = seal(&key(), b"", b"0123456789", &mut random).expect("seals");
        assert_eq!(sealed.len(), 10 + NONCE_LEN + TAG_LEN);
    }

    #[test]
    fn a_wrong_key_fails_to_open() {
        let mut random = SeededRandom::new(2);
        let sealed = seal(&key(), b"ad", b"plaintext", &mut random).expect("seals");
        let other = SymmetricKey::from_bytes([8u8; KEY_LEN]);
        assert_eq!(
            open(&other, b"ad", &sealed),
            Err(CryptoError::DecryptionFailed)
        );
    }

    #[test]
    fn wrong_associated_data_fails_to_open() {
        // This is the property that stops a ciphertext being replayed into a
        // different conversation or a different record.
        let mut random = SeededRandom::new(3);
        let sealed = seal(&key(), b"conversation-a", b"plaintext", &mut random).expect("seals");
        assert_eq!(
            open(&key(), b"conversation-b", &sealed),
            Err(CryptoError::DecryptionFailed)
        );
    }

    #[test]
    fn any_single_bit_flip_is_detected() {
        let mut random = SeededRandom::new(4);
        let sealed = seal(&key(), b"ad", b"plaintext here", &mut random).expect("seals");
        for index in 0..sealed.len() {
            for bit in 0..8u32 {
                let mut tampered = sealed.clone();
                tampered[index] ^= 1 << bit;
                assert!(
                    open(&key(), b"ad", &tampered).is_err(),
                    "a flip at byte {index} bit {bit} went undetected"
                );
            }
        }
    }

    #[test]
    fn a_truncated_message_is_rejected_by_length_not_by_panic() {
        let mut random = SeededRandom::new(5);
        let sealed = seal(&key(), b"", b"x", &mut random).expect("seals");
        for cut in 0..sealed.len() {
            assert!(open(&key(), b"", &sealed[..cut]).is_err());
        }
    }

    #[test]
    fn nonces_differ_between_messages() {
        let mut random = SeededRandom::new(6);
        let first = seal(&key(), b"", b"same plaintext", &mut random).expect("seals");
        let second = seal(&key(), b"", b"same plaintext", &mut random).expect("seals");
        assert_ne!(first[..NONCE_LEN], second[..NONCE_LEN]);
        assert_ne!(
            first, second,
            "identical plaintexts must not produce identical ciphertexts"
        );
    }

    #[test]
    fn an_explicit_nonce_round_trips() {
        let nonce = [3u8; NONCE_LEN];
        let sealed = seal_with_nonce(&key(), &nonce, b"ad", b"body").expect("seals");
        assert_eq!(&sealed[..NONCE_LEN], &nonce);
        let body = &sealed[NONCE_LEN..];
        assert_eq!(
            open_with_nonce(&key(), &nonce, b"ad", body).expect("opens"),
            b"body"
        );
    }

    #[test]
    fn a_key_does_not_print_itself() {
        assert_eq!(format!("{:?}", key()), "SymmetricKey(***)");
    }

    #[test]
    fn a_short_key_is_rejected() {
        assert_eq!(
            SymmetricKey::parse(&[0u8; 16]),
            Err(CryptoError::BadLength {
                what: "symmetric key",
                expected: 32,
                actual: 16
            })
        );
    }
}
